### ISSUE TITLE
Clarify `WebSocket.SendAsync()` limitations and API guarantees

### ISSUE THREAD

**Issue by djee-ms** [dotnet/dotnet-api-docs#8040]
The `WebSocket.SendAsync()` [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket.sendasync?view=net-6.0&source=docs) states that:

> Exactly one send and one receive is supported on each [WebSocket](https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket?view=net-6.0) object in parallel.

However it seems that calling `SendAsync()` multiple times before the previous send completed does not raise any exception, so this statement seems to refer to the internal implementation (the sending is done serially), which obviously comes from a limitation of the WebSocket protocol itself, rather than a limitation in the public API. So the first ask would be to clarify this in the docs, and explain that calling this method multiple times is allowed, and what happens in that case (likely: sends are queued internally).

The second part is, assuming those calls are internally queued, what guarantees does the API provide? Specifically, is the API guaranteeing that sends are executed in the same order as the calls are made? This is critical for some scenario, and currently I found myself having to work around the lack of API guarantee by maintaining my own queue (and therefore, having to copy the data to send into that manual queue), and making calls to `SendAsync()` one by one, waiting for the previous call to complete before I make the next one. This is overly complicated, results in extra copy, and I'm pretty confident is completely useless as I would expect the actual implementation to be doing exactly the same thing internally already (duplicated work).

In summary, this issue is 2 related asks:
1. Clarify whether `SendAsync()` can be called multiple times while a send is pending. This is somewhat related to #7809.
2. If so, clarify the behavior of the implementation in terms of internally enqueuing those send calls, and the ordering guarantee if any.

**Comment by CarnaViire** [comment]
The point is that there's no guarantees for the behavior. Like with non-thread-safe collections in multithreaded scenario -- they may happen to work, but there are no guarantees.

The docs are correct -- only one receive and one send are supported concurrently. The fact that it is currently being ordered, is an implementation detail which may change in the future. The ordering implementation was actually introduced quite recently in .NET 6, and before it was throwing on parallel read and had undefined behavior on parallel write. You can read more expanded explanation here https://github.com/dotnet/runtime/issues/58202#issuecomment-906635617

**Comment by djee-ms** [comment]
> Like with non-thread-safe collections in multithreaded scenario -- they may happen to work, but there are no guarantees.

How are users supposed to ensure message B is sent after message A? If as we suspect this means message B's `SendAsync()` cannot be _called_ (let alone the associated task _completed_) until the task from A's call _completed_, then this should probably be explained in the docs.

```cs
public async Task SendOrdered(A, B) {
    await webSocket.SendAsync(A);
    await webSocket.SendAsync(B);
}

public Task SendUnordered(A, B) {
    Task[] tasks = new(2);
    tasks.Add(webSocket.SendAsync(A));
    tasks.Add(webSocket.SendAsync(B));
    return Task.WhenAll(tasks);
}
```

> The docs are correct -- only one receive and one send are supported concurrently.

This concept doesn't map to API calls. You can call `WebSocket.SendAsync()` multiple times in parallel, so this API does support "multiple sends" in the sense of multiple calls to `SendAsync()`. "One send is supported concurrently" is a conceptual statement about the internal behavior, where the concept of "one send" is not defined. I believe reading your explanation that this means "only one send task will execute at any time, so all tasks from `SendAsync()` calls will be internally serialized (in an undetermined order)". This is what this issue is asking a docs fix for.

**Comment by djee-ms** [comment]
> > How are users supposed to ensure message B is sent after message A?
> 
> Awaiting each call, as in your first example.

Then this should be documented.

> > > The docs are correct -- only one receive and one send are supported concurrently.
> > 
> > This concept doesn't map to API calls. You can call WebSocket.SendAsync() multiple times in parallel, so this API does support "multiple sends" in the sense of multiple calls to SendAsync().
> 
> It does "map" to it, because the term is "supported", not "possible", right? It means if you call the sends serially, i.e. calling each send after awaiting the previous send, we guarantee it will work correctly -- and if it does not, it is a bug we should fix.
> 
> If you call them in parallel, though, you are on your own and we don't guarantee anything. It _might_ happen to work and _might_ happen not to. But it would be neither a bug nor a feature.

Setting aside the playing on word meaning, which doesn't help anyone especially in a public documentation targeted at a wide audience of developers with various backgrounds and level of understanding, so you're saying here that calling `SendAsync()` twice in parallel is _undefined behavior_? This is a way stronger (and worse, usability-wise) statement than what was discussed above. And this should also be detailed. The statement in the docs doesn't merely imply that strong of a limitation.

**Comment by stephentoub** [comment]
@CarnaViire is correct. WebSocket doesn't support multiple concurrent sends nor multiple concurrent receives; such usage is erroneous. What undefined behavior incorrect usage results in, whether an exception or happening to succeed or crashing or anything else, is an implementation detail that can and has changed, and that can and does differ between WebSocket-derived types. If the structure of a developer's code could otherwise result in such concurrent usage (beyond the supported case of one pending send and one pending receive), out is up to the developer to serialize the access via whatever mechanism works best for them. That could be using a semaphore, it could be guaranteeing the previous operation is awaited/conpleted before issuing the next, etc.

**Comment by djee-ms** [comment]
Ok. So please document that undefined behavior which is not at all clear in the docs and is a massive footgun. That the WebSocket _standard_ do not support concurrent send/receive is one (well understood) thing. But that an API returning a `Task` do not support parallel scheduling of such tasks is completely unexpected, and one would reasonably assume all the synchronization you're mentioning @stephentoub to be already implemented internally, even if ordering is not guaranteed, which apparently is not the case.

**Comment by djee-ms** [comment]
...all the more as `SendAsync()` is not throwing any exception if a second call is made in parallel.

**Comment by stephentoub** [comment]
> But that an API returning a Task do not support parallel scheduling of such tasks is completely unexpected,

The vast majority of Task-returning methods are not thread-safe / do not support concurrent use on the same instance.

**Comment by djee-ms** [comment]
> > But that an API returning a Task do not support parallel scheduling of such tasks is completely unexpected,
> 
> The vast majority of Task-returning methods are not thread-safe / do not support concurrent use on the same instance.

This is not what we're talking about. We're not talking about calling `SendAsync()` in parallel from multiple threads. We're talking about calling `SendAsync()` from one thread, and after it returned a `Task`, and before that task completed, calling again `SendAsync()`, even on the same thread. This is not about multithreading, this is about having multiple tasks in flight that did not complete yet.

**Comment by stephentoub** [comment]
The vast majority of Task-returning methods support neither being used in parallel nor concurrently.  For example, it is illegal, whether there's an exception that results or not, to do things like:
```C#
Stream s = ...;
Task t1 = s.ReadAsync(...);
Task t2 = s.ReadAsync(...);
Task.WaitAll(t1, t2);
```

**Comment by djee-ms** [comment]
So this method is not thread-safe, and cannot be called while any other `Task` from a previous call is pending. Why can't we document this properly?

The example with `Stream` makes little sense because it reads from a single stream and advance the read position, so no one in their right mind would ever want to do reads at random positions in a stream. Even if everything was otherwise thread-safe, the API would be unusable.

For WebSocket sending on the other end, each message is sent as a self-contained one, and it makes perfect sense from an app perspective to send messages to a remote endpoint without necessarily needing ordering, if said messages are independent of each other. I don't think the expectation that the calls are internally synchronized and serialized (even without guaranteed ordering) is unreasonable, because this is exactly how most networking "send" APIs work, internally queuing data until actually sent through a physical interface. And the documentation line "Exactly one send and one receive is supported" makes it sound precisely like this is what's happening : we can't send all messages in parallel, so don't expect you will get any bandwidth/speed gain by sending in parallel, as instead we have to serialize those sends into a single socket stream. But it doesn't imply that merely calling `SendAsync()` twice in a row is undefined behavior. I'm just asking we clarify this.
